[FLINK-19339] Support unions with logical types in Avro >= 1.9.x

dawidwys opened the pull request:
## What is the purpose of the change

Avro 1.9.x introduced yet another mechanism for registering/looking up
conversions for logical types.

If a logical type is part of a union a static field MODEL$ of type SpecificData
is added to the generated Avro class with registered conversions for logical types.
In this commit we try to use that SpecificData in AvroSerializer and
Avro(De)SerializationSchema whenever available instead of instantiating a new
unrelated one.

The change is backwards compatible and if the field is not available we
fallback to the old ways.

## Verifying this change

This change added tests in respective test classes for AvroSerializer and Avro(De)SerializationSchema.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (**yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

StephanEwen commented:
Looks good to me.

Is this also relevant for other Avro parts, like AvroInputFormat, AvroSerializer, ... ?

dawidwys commented:
It is relevant for `AvroSerializer` (which I updated through `AvroFactory#fromSpecific` and added a test for it: `AvroUnionLogicalSerializerTest.java`).

I checked all places where we instantiate the `SpecificData`manually. In the `AvroInputFormat` we use the `SpecificDatumReader` which should just work (it has that logic of acquiring `SpecificData` built in).

StephanEwen commented:
Thanks for clarifying!

Looks good to me, +1 to merge this

dawidwys commented:
Thanks @StephanEwen for the review, merging.

AHeise commented:
Seems like this fixed causes a 10% performance regression https://issues.apache.org/jira/browse/FLINK-19439 . @dawidwys would you like to pick that ticket up?
